File: 637_Average_of_Levels_in_Binary_Tree.java
/*
https://leetcode.com/problems/average-of-levels-in-binary-tree/
Given a non-empty binary tree, return the average value of the nodes on each level in the form of an array.

Breadth-first search. Create a list of all non-null nodes at each level.
Time - O(n)
Space - O(n)
*/

class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> result = new ArrayList();

        if (root == null)
            return result;
        List<TreeNode> level = new ArrayList();     // list of nodes at current depth
        level.add(root);

        while (level.size() > 0) {

            long sum = 0;                           // avoid overflow
            List<TreeNode> newLevel = new ArrayList();

            for (TreeNode node : level) {
                sum += node.val;
                if (node.left != null)              // add non-null nodes
                    newLevel.add(node.left);
                if (node.right != null)
                    newLevel.add(node.right);
            }
            result.add((double) sum / (double) level.size());
            level = newLevel;
        }

        return result;
    }
}

File: python_1_to_1000/637_Average_of_Levels_in_Binary_Tree.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/average-of-levels-in-binary-tree/
# Given a non-empty binary tree, return the average value of the nodes on each level in the form of an array.

# Breadth first search. Maintain a queue of nodes in the current level.
# Time - O(n)
# Space - O(n)

class Solution(object):
    def averageOfLevels(self, root):
        """
        :type root: TreeNode
        :rtype: List[float]
        """
        nodes = [root]
        result = []

        while True:

            row_sum, row_count = 0, 0       # no need to use a list to store the values
            new_nodes = []

            for node in nodes:
                if not node:                # ignore None
                    continue
                row_sum += node.val
                row_count += 1
                new_nodes.append(node.left) # add children to next queue
                new_nodes.append(node.right)

            if row_count == 0:
                break
            result.append(row_sum / float(row_count))
            nodes = new_nodes               # check next row even if all nodes are None

        return result